### PR TITLE
TASK: Ban nullable @PropertyDefinitions

### PR DESCRIPTION
**Pull request by ashleyheath**
Require that joda beans @PropertyDefinitions are either validated as notNull or have an optional getter.

### CODE REVIEW

**Comment by jodastephen** [comment]
This doesn't work for primitive fields.

**Comment by ashleyheath** [comment]
Updated to account for primitive properties.

**Comment by wjnicholson** [comment]
Do we care about Optional getters for DynamoDb metadata beans?

**Comment by cjkent** [comment]
> Do we care about Optional getters for DynamoDb metadata beans?

No, those need to be traditional beans with setters and getters that can return null.

**Comment by ashleyheath** [comment]
I see two different ways of addressing that:
1. Set checkstyle off for the dynamo files (with comments explaining null values are appropriate here
2. Change the regex to allow an explicit empty string in the `@PropertyDefinition`, i.e. `validate = ""`

**Comment by jodastephen** [comment]
There seem to be quite a few nullable fields in Eurex too so this will be painful to introduce. I guess option 2 works, as I don't think turning checkstyle on and off is terribly appealing.

**Comment by ashleyheath** [comment]
Have tweaked the regex to allow `@PropertyDefinition(validate = "")`, so if you want a nullable property you have to explicitly make (and document) that decision.

**Comment by ashleyheath** [comment]
Any objections to that or am I good to merge this in?

File: build-config/src/main/resources/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -190,13 +190,14 @@
...
       <property name="illegalPattern" value="true"/>
     </module>
-    <module name="Regexp">
-      <!-- Matching closing argument e.g. `String foo, String bar) {` -->
-      <!-- Note: Does not match multiple arguments on their own line e.g. `String foo, String bar,` -->
-      <property name="format" value="^[^(\r\n]+[,](?![^,]*>)[^(\n\r]+\)[ ]*\{[ ]*$"/>
-      <property name="message" value="When split across multiple lines, each parameter must be placed on its own line"/>
-      <property name="illegalPattern" value="true"/>
-    </module>
+<!--    Regex incorrectly matches method parameters with multiple type params.-->
```
Comment (wjnicholson) [inline build-config/src/main/resources/checkstyle/checkstyle.xml]: Can this not be a separate PR?

File: build-config/src/main/resources/checkstyle/checkstyle.xml [hunk 2]
```diff
@@ -207,6 +208,11 @@
       <property name="message" value="Closing parenthesis should be on the same line as the last parameter"/>
       <property name="illegalPattern" value="true"/>
     </module>
+    <module name="Regexp">
+      <property name="format" value="@PropertyDefinition(?!(.*validate = |.*get = \&quot;optional&quot;))[\S ]*(\n|\r\n)(?![\S ]* (byte|char|short|int|long|float|double|boolean) [\S ]*$)"/>
+      <property name="message" value="Properties should either be declared 'validate = &quot;notNull&quot;' or 'get = &quot;optional&quot;'"/>
```
Comment (wjnicholson) [inline build-config/src/main/resources/checkstyle/checkstyle.xml]: So does this match any property definition with a `validate = ` section?
Looking and it seems there's a lot of code where this is going to need to be fixed, spread across all projects. Are you sure you want to do that update?

File: build-config/src/main/resources/checkstyle/checkstyle.xml (hunk 1 above)
Comment (ashleyheath) [reply build-config/src/main/resources/checkstyle/checkstyle.xml]: Have just removed it entirely; too painful to get working correctly.

File: build-config/src/main/resources/checkstyle/checkstyle.xml (hunk 2 above)
Comment (ashleyheath) [reply build-config/src/main/resources/checkstyle/checkstyle.xml]: I don't think we should shy away from fixing code that doesn't conform to the standards. I'm more concerned with getting conformation that the rule change appropriately captures those standards.

**Comment by jodastephen** [comment]
Merging this, but testing shows two other problems with recent changes so far:

"Regexp: Line matches the illegal pattern 'When split across multiple lines, method parameters must be followed by a new line'."
```
    if (calculationTarget instanceof BulletPaymentTrade ||
        calculationTarget instanceof CdsTrade ||
        calculationTarget instanceof CdsIndexTrade) {
      return CurveMapping.DEFAULT;
    }
```

"Regexp: Line matches the illegal pattern 'When split across multiple lines, method parameters must be followed by a new line'."
```
  private static ValueWithFailures<Map<ObservableId, LocalDateDoubleTimeSeries>> parse(
      CharSource charSource,
      String location) {
    // Implementation node:
````
